### ISSUE TITLE
Mesh.classify_position: sidedness from the nearest triangle CENTROID misreads points just outside an end-cap rim

### ISSUE THREAD

**Issue by rutgerfick** [dmrai-lab/dmipy-sim#20]
## Symptom

`classify_position` reports **interior** for points that are outside a closed mesh, in a thin shell around
an end-cap rim. Because `init_positions` keeps whatever it labels interior, those points get seeded as
"intra" walkers and then diffuse in open space.

Measured on a closed tube (`trimesh.creation.cylinder(radius=0.4, height=40.0, sections=24)`, watertight,
volume 19.877 µm³ vs πr²h = 20.106) in a 16 × 16 × 44 µm box with `feature_radius=0.4e-6`, seeding 1500
intra walkers:

| | count |
|---|---|
| seeds the mesh calls interior | 1500 |
| actually inside (`trimesh.contains`) | 1409 |
| **misclassified** | **91** |

All 91 have r > 0.400 µm, i.e. beyond the circumradius — this is not polygon-vs-circle ambiguity (0 of them
fall in the inradius band 0.3966–0.400), and not surface adjacency (59 at r 0.45–0.8, 25 at r 0.8–1.2).

**They are concentrated at the caps: 98.9% lie within 1 µm of an end cap** (61 at |z| 19.0–19.9, 29 at
19.9–20.0, half-length 20.0).

Away from the caps classification is exact. Radial probes at mid-length read correctly at
r = 0.10, 0.30, 0.39 (interior) and 0.45, 0.6, 0.9, 1.2, 2.0, 4.0 µm (exterior). Axial probes:

```
r=0.8  z=10.00  -> exterior  (correct)
r=0.8  z=19.50  -> exterior  (correct)
r=0.8  z=19.90  -> exterior  (correct)
r=0.6  z=19.99  -> INTERIOR  (wrong)
```

## Cause

Sidedness is taken from the nearest triangle **centroid**:

```python
dist = jnp.where(valid, jnp.linalg.norm(r_w[None] - cent, axis=1), jnp.inf)
idx  = jnp.argmin(dist)
side = jnp.dot(r_w - cent[idx], nrm[idx])
```

Just outside the rim, the nearest *centroid* belongs to an **end-cap** facet whose normal points along ±z.
For a query point axially below the cap but radially outside the wall, `dot(r − cent, n_z) < 0`, which reads
as interior. The nearest *centroid* is not the nearest *surface point*, and near a convex edge the two pick
different facets.

The `valid.any()` guard added in #21 does not apply here: triangles **are** present in the gather, so this
is the sidedness heuristic, not the empty-space fallback.

## Scope

- Present in both the public and private engines (same 1409/1500 on the same tube and seed), so it predates
  the guard and is not a regression.
- The 6% rate is **specific to that test geometry** — a 40 µm tube in a 16 × 16 µm box puts a large share of
  accepted points in the rim shell. It is a rim artifact, not a uniform error rate. A longer substrate
  dilutes it; a short or highly branched one concentrates it.
- **Where it matters:** substrates that are *cut* rather than closed-by-construction, e.g. the XNH axon
  meshes, which are clipped at both ends — exactly where this fires. Worth quantifying on a real axon before
  a production walk.

## Possible fixes

1. **Nearest point on the triangle** instead of the centroid for the sidedness test (point-triangle distance
   with barycentric clamping), then the normal of that facet. Correct at convex edges; costs a little more
   arithmetic per candidate.
2. **Winding number / ray parity** over the gathered candidates — robust for any closed surface, more work
   per query.
3. **Signed distance from the k nearest facets**, taking the sign of the smallest |distance| rather than of
   the nearest centroid.

(1) is probably the best cost/benefit, and it reuses the existing candidate gather.

## Repro

```python
import numpy as np, trimesh, jax
from dmipy_sim.mesh import Mesh
tube = trimesh.creation.cylinder(radius=0.4, height=40.0, sections=24)
V = np.asarray(tube.vertices, float)*1e-6; F = np.asarray(tube.faces, np.int64)
m = Mesh(V, F, periodic=False, voxel_min=np.array([-8e-6,-8e-6,-22e-6]),
         voxel_max=np.array([8e-6,8e-6,22e-6]), feature_radius=0.4e-6)
pts = np.asarray(m.init_positions(1500, jax.random.PRNGKey(0), intra=True))
ins = tube.contains(pts*1e6)
print(ins.sum(), "of", len(pts), "actually inside")
z = np.abs(pts[~ins, 2])*1e6
print(f"{100*(z>19).mean():.1f}% of the bad seeds are within 1 um of an end cap")
```

**Comment by rutgerfick** [comment]
**Closing this: the premise was wrong, and it was my own.**

I raised this from a test tube built as `trimesh.creation.cylinder(radius=0.4, height=40.0, sections=24)` —
**96 faces**, whose side triangles span the entire 40 µm length. On a mesh like that the nearest *centroid*
can be 6.7 µm from a query point whose nearest *surface* is 0.6 µm away, which is what produced the 6%
misclassification I reported.

That geometry is pathological, not representative. Measured properly:

| mesh | faces | median edge | agreement with ray casting |
|---|---|---|---|
| coarse tube (my test) | 96 | 0.400 | 72.70% |
| subdivided tube | 6,144 | 0.050 | **100.00%** |
| real XNH axon (axon06 inner) | 159,154 | 0.137 | **99.80%** (6 false-inside / 3000, 0 false-outside) |

I also implemented the proposed fix — sidedness from the closest *point* on the closest of the k nearest
candidate triangles (barycentric clamping) instead of the nearest centroid — and it is a **no-op** on both the
fine tube and the real axon: bit-identical verdicts to the existing code in every case. It only differs on the
degenerate mesh, which no substrate here resembles.

So there is no defect to fix for reasonably tessellated meshes, and the existing centroid test is fine at the
triangle scales these substrates use. Not merging the change: it adds a barycentric inner loop and a k-nearest
gather for no measured gain.

Worth recording for anyone who revisits it: if a future substrate does carry facets much larger than the
feature being resolved, this becomes real again, and the nearest-point variant is the fix. The branch
`fix/nearest-point-sidedness` has a working implementation if it is ever needed.
